harness/determined/common/context.py

```diff
@@ -66,34 +66,58 @@ def v1File_from_local_dir(path: str, local_path: pathlib.Path) -> bindings.v1Fil
 
 
 def read_v1_context(
-    local_path: pathlib.Path,
+    root_path: pathlib.Path,
     limit: int = constants.MAX_CONTEXT_SIZE,
 ) -> List[bindings.v1File]:
     """
     Return a list of v1Files suitable for submitting a context directory over the v1 REST API.
 
+    If root_path refers to a file, a context directory is created containing only that file.
+
     A .detignore file in the directory, if specified, indicates the wildcard paths
     that should be ignored. File paths are represented as relative paths (relative to
     the root directory).
     """
 
+    root_path = root_path.resolve()
+
+    if not root_path.exists():
+        raise Exception(f"Path '{root_path}' doesn't exist")
+
     items = []
     size = 0
 
     def add_item(f: bindings.v1File) -> None:
         nonlocal size
```

Comment:
Quibble, but all of the `nonlocal` calls seem awkward to me.  Probably not worth the re-write, but I'd personally find the code easier to follow if, for instance, `add_item` took on a `size` arg and returned the updated size:
```python
def add_item(f: bindings.v1File, size: int) -> int:
    # ....
    size += v1File_size(f)
    # ...
    return size
```
and use the output to update `size` in the outer scope.  Similar for `msg`

Reply:
The level at which I care about code cleanliness changes according to the scope it affects.

Changes that affect a whole code base are really important to me, and I care a lot about even very tiny details.  Changes which only affect a single file are of small importance.  Changes that only affect the inside of a single function are firmly in the "don't care" category, even though I don't disagree with your reasoning... I like when we can write functions without side effects.

In this particular case though, I would also point out that even if you avoided the `nonlocal size` you'd still be modifying `items` inside `add_item()` so returning size explicitly doesn't actually make the function pure.  I personally find that _more_ confusing.

Reply:
All fair. As I said, a quibble, and a good point about `items`.